src/main/java/com/google/cloud/flink/bigquery/BigQuerySourceFunction.java

```diff
@@ -67,77 +71,70 @@ public TypeInformation<RowData> getProducedType() {
 
   @Override
   public void open(Configuration parameters) throws Exception {
-
-    this.executorIndex = getRuntimeContext().getIndexOfThisSubtask();
+    deserializer.open(() -> getRuntimeContext().getMetricGroup().addGroup("bigQuery"));
+    this.executerIndex = getRuntimeContext().getIndexOfThisSubtask();
     this.numOfExecutors = getRuntimeContext().getNumberOfParallelSubtasks();
     this.numOfStreams = readSessionStreamList.size();
-    for (int i = executorIndex; i < numOfStreams; i += numOfExecutors) {
+    this.streamNames.clear();
+    for (int i = executerIndex; i < numOfStreams; i += numOfExecutors) {
       if (running) {
         this.streamNames.add(readSessionStreamList.get(i));
       }
     }
   }
 
+  @SuppressWarnings("resource")
   @Override
-  public void run(final SourceContext<RowData> ctx) throws Exception {
-    Collector<RowData> collector = new SourceContextCollector<RowData>(ctx);
+  public void run(SourceContext<RowData> ctx) throws Exception {
+    List<RowData> outputCollector = new ArrayList<>();
```

Comment:
Why the outputcollector is needed? can't we run ctx.collect() directly on the parsed data?

Reply:
@davidrabinowitz , context.collect() receives single row data only, Since we are processing set of records from the read response, we are collecting all the converted records to 'outputcollector', then adding to context one by one.